sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,265 @@
+
+    std::vector<bool> final_valid;
+    std::vector<uint64_t> final_iaddr;
+    for (unsigned i = 0; i < tracerv_width; i++) {
+      final_valid.emplace_back(0);
+      final_iaddr.emplace_back(0xffff);
+    }
+
+    return std::make_pair(final_iaddr, final_valid);
+  }
+
+  // call init on tracerV
+  // call tick 1:1 with take_steps
+  // copy termination brige take_steps
+  // call take_steps for 100
+  // the in a loop, tick all bridges, untill done() is true
+  //
+
+  // expose a vec of iaddr, and valid
+  //   consider ValidIO,, a bundle with valid implicit (valid/ready without the
+  //   ready) just expose iaddr
+
+  // add require to traverVBridge
+  // consider PR a test before widht changes
+
+  // take N steps
+  // tick all of the bridges until done is asserted
+  // there is a timeout to see if done never returns
+  // returns false for error
+  // returns true for success
+  bool steps(const unsigned s) {
+    simif->take_steps(s, /*blocking=*/false);
+    const unsigned timeout = 10000 + s;
+    bool was_done = false;
+    for (unsigned i = 0; i < timeout; i++) {
+
+      for (auto &bridge : bridges) {
+        bridge->tick();
+      }
+      tracerv->tick();
+
+      if (simif->done()) {
+        was_done = true;
+        break;
+      }
+    }
+
+    if (!was_done) {
+      std::cout << "Hit timeout of " << timeout
+                << " tick loops afer a requested " << s << " steps"
+                << std::endl;
+    }
+
+    return was_done;
+  }
+
+  std::vector<uint64_t> expected_cycle;
+  std::vector<uint64_t> expected_iaddr;
+
+  std::vector<uint64_t> got_cycle;
+  std::vector<uint64_t> got_iaddr;
+
+  void got_instruction(const uint64_t cycle, const uint64_t pc) {
+    // std::cout << "cycle: " << cycle << " pc: " << pc << std::endl;
+    got_cycle.emplace_back(cycle);
+    got_iaddr.emplace_back(pc);
+  }
+
+  int simulation_run() override {
+    if (!tracerv) {
+      std::cout << "tracerv was never set" << std::endl;
+    }
+
+    // set the callback to capture traced instructions
+    tracerv->set_callback(std::bind(&TracerVModule::got_instruction,
+                                    this,
+                                    std::placeholders::_1,
+                                    std::placeholders::_2));
+
+    // Reset the DUT.
+    peek_poke->poke("reset", 1, /*blocking=*/true);
+    simif->take_steps(1, /*blocking=*/true);
+    peek_poke->poke("reset", 0, /*blocking=*/true);
+    simif->take_steps(1, /*blocking=*/true);
+
+    // the value of the first cycle as returned from TracerV
+    const uint64_t cycle_offset = 3;
+
+    // modified as we go
+    uint64_t e_cycle = cycle_offset;
+
+    const unsigned tracerv_width = tracerv->max_core_ipc;
+
+    // load MMIO and capture expected outputs
+    auto load = [&](std::vector<uint64_t> iad, std::vector<bool> bit) {
+      assert(iad.size() == bit.size());
+      const auto sz = iad.size();
+      for (unsigned i = 0; i < sz; i++) {
```

Comment:
nit: In LLVM, we usually write: `for (unsigned i = 0, sz = iad.size(); i < sz; ++i)`